File: Day3/part2.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
#include <climits>
using namespace std;

#define N 1500
#define FASTIO                        \
    ios_base::sync_with_stdio(false); \
    cin.tie(0);                       \
    cout.tie(0);
#define x first
#define y second
typedef long long ll;
typedef pair<int, int> pii;

int rows, columns;
string grid[N];

int findSlope(int right, int down)
{
    int col = right;
    int count = 0;

    for (int i = down; i < rows; i += down)
    {
        if (grid[i][col] == '#')
        {
            count++;
        }

        col = (col + right) % columns;
    }

    return count;
}

int main()
{
    FASTIO;

    rows = 0;
    while (cin >> grid[rows])
    {
        rows++;
    }

    columns = grid[0].size();

    ll ans = 1;
    vector<pii> moves = {{1, 1}, {3, 1}, {5, 1}, {7, 1}, {1, 2}};
    for (auto move : moves)
    {
        cout << move.x << " " << move.y << " " << findSlope(move.x, move.y) << endl;
        ans *= findSlope(move.x, move.y);
    }

    cout << ans << endl;

    return 0;
}

File: python/Day3/part2.py
#!/usr/bin/env python3
from sys import stdin

grid = stdin.readlines()
rows = len(grid)
columns = len(grid[0]) - 1  # newline


def findSlope(right, down):
    col = right
    count = 0

    for i in range(down, rows, down):
        if grid[i][col] == "#":
            count += 1

        col = (col + right) % columns

    return count


ans = 1
moves = ((1, 1), (3, 1), (5, 1), (7, 1), (1, 2))
for move in moves:
    ans *= findSlope(*move)

print(ans)
